File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_10/ControlFlowFirstTest.java
package com.fResult.reactor.ch5_10;

import java.time.Duration;
import lombok.extern.log4j.Log4j2;
import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

@Log4j2
class ControlFlowFirstTest {
  @Test
  void first() {
    val slow = Flux.just(1, 2, 3).delayElements(Duration.ofMillis(10));
    val fast = Flux.just(4, 5, 6, 7).delayElements(Duration.ofMillis(1));
    val first = Flux.firstWithSignal(slow, fast);

    StepVerifier.create(first).expectNext(4, 5, 6, 7).verifyComplete();
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_10/ControlFlowFirstTest.kt
package com.fResult.reactor.ch5_10

import reactor.core.publisher.Flux
import reactor.test.StepVerifier
import kotlin.test.Test
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.toJavaDuration

class ControlFlowFirstTest {
  @Test
  fun first() {
    val slow = Flux.just(1, 2, 3).delayElements(10.milliseconds.toJavaDuration())
    val fast = Flux.just(4, 5, 6, 7).delayElements(2.milliseconds.toJavaDuration())
    val firstEmittedStream = Flux.firstWithSignal(slow, fast)

    StepVerifier.create(firstEmittedStream).expectNext(4, 5, 6, 7).verifyComplete()
  }
}
